pkg/virtual/workspaces/builder/org_listener.go

```diff
@@ -0,0 +1,180 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package builder
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/authentication/user"
+
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1/helper"
+	workspaceauth "github.com/kcp-dev/kcp/pkg/virtual/workspaces/auth"
+	workspacecache "github.com/kcp-dev/kcp/pkg/virtual/workspaces/cache"
+	virtualworkspacesregistry "github.com/kcp-dev/kcp/pkg/virtual/workspaces/registry"
+)
+
+// orgListener listens for changes in the root WorkspaceAuthCache,
+// in order to manage the list of orgs
+type orgListener struct {
+	newOrg func(orgName string) virtualworkspacesregistry.Org
+
+	clusterWorkspaceCache *workspacecache.ClusterWorkspaceCache
+
+	knownWorkspaces   sets.String
+	orgListerUserInfo user.Info
+
+	ready    func() bool
+	orgs     map[string]virtualworkspacesregistry.Org
+	orgMutex sync.RWMutex
+}
+
+func NewOrgListener(orgListerUserInfo user.Info, clusterWorkspaceCache *workspacecache.ClusterWorkspaceCache, rootOrg virtualworkspacesregistry.Org, newOrg func(orgName string) virtualworkspacesregistry.Org) *orgListener {
+	w := &orgListener{
+		orgListerUserInfo: orgListerUserInfo,
+		newOrg:            newOrg,
+		orgs: map[string]virtualworkspacesregistry.Org{
+			helper.RootCluster: rootOrg,
+		},
+
+		clusterWorkspaceCache: clusterWorkspaceCache,
+
+		knownWorkspaces: sets.NewString(),
+
+		ready: func() bool { return false },
+	}
+	return w
+}
+
+func (l *orgListener) Initialize(authCache *workspaceauth.AuthorizationCache) {
+	l.orgMutex.Lock()
+	defer l.orgMutex.Unlock()
+
+	var mutex sync.Mutex
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	readys := []func() bool{}
+	workspaces, _ := authCache.List(l.orgListerUserInfo, labels.Everything())
+	for _, workspace := range workspaces.Items {
+		orgName := helper.RootCluster + ":" + workspace.Name
+		org := l.newOrg(orgName)
+		readys = append(readys, org.ReadyForAccess)
+		l.knownWorkspaces.Insert(workspace.Name)
+		l.orgs[orgName] = org
+	}
+	l.ready = func() bool {
+		mutex.Lock()
+		defer mutex.Unlock()
+
+		for _, r := range readys {
+			if !r() {
+				return false
+			}
+		}
+		return true
+	}
+	authCache.AddWatcher(l)
+}
+
+func (l *orgListener) Stop() {
+	l.orgMutex.RLock()
+	defer l.orgMutex.RUnlock()
+
+	for _, o := range l.orgs {
+		o.Stop()
+	}
+}
+
+func (l *orgListener) GroupMembershipChanged(workspaceName string, users, groups sets.String) {
+	orgName := helper.RootCluster + ":" + workspaceName
```

Comment:
use helper

Reply:
Fixed in commit c48becc80ef541475a0e032a26cfa6a9024c5ac8